svm/src/account_loader.rs

```diff
@@ -95,6 +95,93 @@ pub struct FeesOnlyTransaction {
     pub fee_details: FeeDetails,
 }
 
+#[cfg_attr(feature = "dev-context-only-utils", derive(Clone))]
+pub(crate) struct AccountLoader<'a, CB: TransactionProcessingCallback> {
+    pub(crate) program_cache: ProgramCacheForTxBatch,
+    account_overrides: Option<&'a AccountOverrides>,
+    callbacks: &'a CB,
+}
+impl<'a, CB: TransactionProcessingCallback> AccountLoader<'a, CB> {
+    pub fn new(
+        callbacks: &'a CB,
+        account_overrides: Option<&'a AccountOverrides>,
+        program_cache: ProgramCacheForTxBatch,
+    ) -> AccountLoader<'a, CB> {
+        Self {
+            program_cache,
+            account_overrides,
+            callbacks,
+        }
+    }
+
+    pub fn load_account(
+        &mut self,
+        account_key: &Pubkey,
+        usage_pattern: AccountUsagePattern,
+    ) -> Option<LoadedTransactionAccount> {
+        let is_writable = usage_pattern == AccountUsagePattern::Writable;
+        let is_invisible_read = usage_pattern == AccountUsagePattern::ReadOnlyInvisible;
+
+        if let Some(account_override) = self
+            .account_overrides
+            .and_then(|overrides| overrides.get(account_key))
+        {
+            Some(LoadedTransactionAccount {
+                loaded_size: account_override.data().len(),
+                account: account_override.clone(),
+                rent_collected: 0,
+            })
+        } else if let Some(program) = is_invisible_read
+            .then_some(())
+            .and_then(|_| self.program_cache.find(account_key))
+        {
+            self.callbacks.get_account_shared_data(account_key)?;
+            // Optimization to skip loading of accounts which are only used as
+            // programs in top-level instructions and not passed as instruction accounts.
+            Some(LoadedTransactionAccount {
+                loaded_size: program.account_size,
+                account: account_shared_data_from_program(&program),
+                rent_collected: 0,
+            })
+        } else if let Some(account) = self.callbacks.get_account_shared_data(account_key) {
```

Comment:
once i rebase on alexander, i am removing this `get_account_shared_data()` call:
* its wrong. the changes to the flow means if it were ever it account loading would believe there is a new account to create rather than aborting
* it is impossible to ever trigger because program cache is built directly from accounts-db

alexander is removing the test that artificially triggers this condition